client/blocks/qr-code-login/index.jsx

```diff
@@ -0,0 +1,185 @@
+
+const getLoginActionResponse = async ( action, args ) => {
+	const url = new URL( 'https://wordpress.com/wp-login.php' );
+	url.searchParams.append( 'action', action );
+
+	Object.keys( args ).forEach( ( key ) => {
+		url.searchParams.append( key, args[ key ] );
+	} );
+
+	const response = await fetch( url.href );
+	return await response.json();
+};
+
+const fetchQRCodeData = async ( setTokenData, tokenData, anonymousUserId ) => {
+	if ( isStillValidToken( tokenData ) ) {
+		return tokenData;
+	}
+	// tokenData is set to null initially.
+	// Lets wait till it is set to false when the local data is the just yet.
+	if ( tokenData === null ) {
+		return null;
+	}
+
+	const responseData = await getLoginActionResponse( 'qr-code-token-request-endpoint', {
+		anon_id: anonymousUserId,
+	} );
+
+	setTokenData( responseData.data );
+	setStoredItem( LOCALE_STORAGE_KEY, responseData.data );
+};
+
+const fetchAuthState = async ( setAuthState, setTokenData, tokenData, anonymousUserId ) => {
+	if ( ! tokenData ) {
+		return;
+	}
+
+	if ( ! isStillValidToken( tokenData ) ) {
+		fetchQRCodeData( setTokenData, tokenData, anonymousUserId );
+		return;
+	}
+
+	const { token, encrypted } = tokenData;
+
+	const responseData = await getLoginActionResponse( 'qr-code-authentication-endpoint', {
+		anon_id: anonymousUserId,
+		token: token,
+		data: encrypted,
+	} );
+
+	setAuthState( responseData.data );
+};
+
+function TokenQRCode( { tokenData } ) {
+	if ( ! tokenData ) {
+		return <QRCodePlaceholder />;
+	}
+	const { token, encrypted } = tokenData;
+	const imageSettings = {
+		src: qrCenter,
+		height: 64,
+		width: 64,
+		excavate: false,
+	};
+	return (
+		<QRCode
+			value={ `https://apps.wordpress.com/get/?campaign=login-qr-code#qr-code-login?token=${ token }&data=${ encrypted }` }
+			size={ 300 }
+			imageSettings={ imageSettings }
+		/>
+	);
+}
+
+function QRCodePlaceholder() {
+	return (
+		<div className="qr-code-login__placeholder">
+			<span className="qr-code-login__corner-box"></span>
+			<span className="qr-code-login__corner-box"></span>
+			<span className="qr-code-login__corner-box"></span>
+		</div>
+	);
+}
+
+function QRCodeLogin() {
+	const translate = useTranslate();
+	const [ tokenData, setTokenData ] = useState( null );
+	const [ authState, setAuthState ] = useState( false );
+
+	const anonymousUserId = getTracksAnonymousUserId();
+
+	// Fetch QR code data.
+	useEffect( () => {
+		fetchQRCodeData( setTokenData, tokenData, anonymousUserId );
+	}, [ setTokenData, tokenData, anonymousUserId ] );
+
+	// Fetch the Auth Data.
+	useInterval( () => {
+		fetchAuthState( setAuthState, setTokenData, tokenData, anonymousUserId );
+	}, AUTH_PULL_INTERVAL );
+
+	useEffect( () => {
+		if ( authState?.auth_url ) {
+			window.location.replace( authState.auth_url );
+		}
+	}, [ authState ] );
+
+	useEffect( () => {
+		getStoredItem( LOCALE_STORAGE_KEY ).then( ( localTokenData ) =>
+			setTokenData( localTokenData ?? false )
+		);
+	}, [] );
+
+	const steps = [
```

Comment:
It's a bummer that we declare the steps on every rerender. This definitely offers some opportunities for improvement. For example, we could wrap these in a `useMemo()` hook call, where they'd re-render only if the locale changes.